### PR TITLE
Add ragger navigator features

### PR DESCRIPTION
**Pull request by xchapron-ledger**

### CODE REVIEW

File: src/ragger/backend/interface.py [hunk 1]
```diff
@@ -287,6 +293,24 @@ def both_click(self) -> None:
...
+    def navigate(self, instruction: NavigationInstruction) -> None:
+        """
+        Performs a click on buttons of the device depending of the navigation instruction.
+
+        :raises NotImplementedError: If the navigation instruction is not implemented.
+
+        :return: None
+        :rtype: NoneType
+        """
+        if instruction == NavigationInstruction.GO_TO_NEXT_SCREEN:
```
Comment (agrojean-ledger) [inline src/ragger/backend/interface.py:305]: I don't understand why is this code in the interface definition

Comment (xchapron-ledger) [reply src/ragger/backend/interface.py:305]: because it is common to every backend? At least as long as we keep `right_click`, `left_click` and `both_click`?

File: src/ragger/backend/speculos.py [hunk 2]
```diff
@@ -177,27 +183,27 @@ def navigate_until_snap(self,
                             start_img_idx: int = 0,
                             last_img_idx: int = 0,
                             take_snaps: bool = True,
-                            timeout: int = 30,
+                            timeout: int = 5,
```
Comment (agrojean-ledger) [inline src/ragger/backend/speculos.py:186]: 5 seconds is too short a timeout if there are many screens in a flow, it will result in tests failing when they should succeed

File: src/ragger/backend/speculos.py [hunk 3]
```diff
@@ -139,34 +149,30 @@ def _screenshot_equal_with_crop(self, path: Path, bytes: BytesIO, crop: Crop = N
...
             screenshot = self._client.get_screenshot()
-            if self._screenshot_equal_with_crop(path, BytesIO(screenshot), crop):
+            if self._snap_equal_with_crop(path, BytesIO(screenshot), crop):
                 return True
             now = time()
         return False
 
     def _compare_snaps(self, path: Path, test_case_name: Path, last_img_idx: int = 0) -> bool:
 
-        snapshot_golden_path = path / "snapshots" / self._firmware.device / test_case_name
```
Comment (agrojean-ledger) [inline src/ragger/backend/speculos.py:157]: Why the hassle of renaming these variables just to make them a bit shorter ?

File: src/ragger/backend/interface.py [hunk 4]
```diff
@@ -359,3 +383,28 @@ def navigate_until_snap(self,
         :rtype: NoneType
         """
         raise NotImplementedError
+
+    @abstractmethod
+    def navigate_and_compare(self, path: Path, test_case_name: Path,
+                             instructions: List[NavigationInstruction]) -> bool:
+        """
+        Navigate and compare snapshots.
```
Comment (agrojean-ledger) [inline src/ragger/backend/interface.py:391]: Make it explicit that there are as many navigation steps as provided by the instructions parameter

File: src/ragger/backend/speculos.py (hunk 3 above)
Comment (xchapron-ledger) [reply src/ragger/backend/speculos.py:157]: the idea is that we have `screenshot`, `snap` and `snapshot` naming, I think it would be better to harmonize on one name?

File: README.md [hunk 5]
```diff
@@ -65,8 +65,10 @@ The `src/ragger/backend/interface.py` file describes the methods that can be imp
 * `right_click`: perform a right click on a device.
 * `left_click`: perform a left click on a device.
 * `both_click`: perform a click on both buttons (left + right) of a device.
+* `navigate`: performs a click on buttons of the device depending of the navigation instruction.
```
Comment (agrojean-ledger) [inline README.md:68]: I would reword it a bit : 

`navigate` : navigate on the device according to a set of navigation instructions provided.
`navigate and compare` : same as the previous method but take screenshots at every step and compare them with "golden images" after navigation is performed.

File: tests/backend/test_speculos.py [hunk 6]
```diff
@@ -172,6 +173,22 @@ def test_navigate_and_compare_until_snap(self):
                         ROOT_SCREENSHOT_PATH, "generic", 0, 2)
                 self.assertTrue(ret)
 
+    def test_navigate_and_compare(self):
```
Comment (agrojean-ledger) [inline tests/backend/test_speculos.py:176]: Not necessary to test non-nominal cases ?

File: src/ragger/backend/interface.py (hunk 1 above)
Comment (agrojean-ledger) [inline src/ragger/backend/interface.py:305]: @lpascal-ledger do you think this should be implemented in this way (in the interface directly) ?

File: src/ragger/backend/speculos.py (hunk 2 above)
Comment (lpascal-ledger) [reply src/ragger/backend/speculos.py:186]: `timeout` could be a one-image parameter, which would be extended when another screenshot is encountered. That would bring more flexibility.

Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:186]: Not sure what you mean :thinking:

Comment (xchapron-ledger) [reply src/ragger/backend/speculos.py:186]: That makes very long `ragger` test, but ok. I'm switching back to 30sec.

File: tests/backend/test_speculos.py (hunk 6 above)
Comment (xchapron-ledger) [reply tests/backend/test_speculos.py]: tests added :)

File: src/ragger/backend/interface.py (hunk 1 above)
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: @xchapron-ledger actually I think the navigation method should be given a list of instructions and loop through them. Additionnaly, like navigate until snap , a "takesnaps" boolean parameter should be given to decide whether or not to take snapshots or not at every step.

Then , navigate and compare would just pass its own navigation instructions to navigate.

File: src/ragger/backend/speculos.py [hunk 7]
```diff
@@ -248,3 +250,20 @@ def navigate_and_compare_until_snap(self,
...
         # Then compare all snapshots taken to golden images.
         return self._compare_snaps(path, test_case_name, img_idx)
+
+    def navigate_and_compare(self, path: Path, test_case_name: Path,
+                             instructions: List[NavigationInstruction]) -> bool:
+        snaps_tmp_path = self._get_snaps_dir_path(path, test_case_name, False)
+        snaps_tmp_path.mkdir(parents=True, exist_ok=True)
+
+        # First navigate to the last step and take snapshots of every screen in the flow.
+        for idx, instruction in enumerate(instructions):
```
Comment (agrojean-ledger) [inline src/ragger/backend/speculos.py:260]: The navigate method should rather have a take snap parameter that would be set to true when called from navigate_and_compare rather than have a loop here and pass each instruction to another loop in navigate...

Comment (xchapron-ledger) [reply src/ragger/backend/speculos.py:260]: Humf, I'm not found of this one. This would need to pass many parameters to the `navigate` method (snaps dir path, test case name, snap index, ...).
I rather go back and allow only one navigation instruction per `navigate` call.
What do you think?

Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:260]: Then you might as well make the navigate method private if it's only used for one navigation instruction (the user can directly use "click_right" "click_left" and so on, I think).

File: src/ragger/backend/interface.py [hunk 8]
```diff
@@ -29,6 +29,12 @@ class RaisePolicy(Enum):
     RAISE_ALL = auto()
 
 
+class NavigationInstruction(Enum):
```
[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/ragger/backend/interface.py:33]: Given navigation will probably be a thing on its own, I'd say it could deserve its own module.

[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/ragger/backend/interface.py:33]: But currently it is entangled with one of the concrete method of the interface, so it would make sense _for now_ to keep it inside the `backend` package.

File: src/ragger/backend/interface.py (hunk 1 above)
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: Ok

File: src/ragger/backend/interface.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:33]: @xchapron-ledger ?

**Review by agrojean-ledger — DISMISSED** [review]

File: src/ragger/backend/interface.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:33]: Ok, I close this discussion because the future refactoring will probably throw this out of the windows.

File: src/ragger/backend/interface.py [hunk 9]
```diff
@@ -362,3 +390,29 @@ def navigate_until_snap(self,
         :rtype: NoneType
         """
         raise NotImplementedError
+
+    @abstractmethod
+    def navigate_and_compare(self, path: Path, test_case_name: Path,
```
Comment (fbeutin-ledger) [inline src/ragger/backend/interface.py:395]: En plusieurs lignes c'est plus lisible

**Review by fbeutin-ledger — DISMISSED** [review]

File: src/ragger/navigator/interface.py [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+        :type middle_instruction_wait: float
+        :param last_instruction_wait: Sleeping time after the last navigation instruction
+        :type last_instruction_wait: float
+
+        :raises ValueError: If one of the snapshots do not match.
+
+        :return: None
+        :rtype: NoneType
+        """
+        if not isinstance(self._backend, SpeculosBackend):
```
Comment (lpascal-ledger) [inline src/ragger/navigator/interface.py:159]: Not found of this. The whole point of the backends is to avoid these discriminated treatments. if we start splitting the usage depending on the backend type, the code will soon become illegible condition hell.
I'd rather accept that code run uselessly - but won't crash, all methods should be defined and correct - on some backend. **For now**. At some point, we will need those empty functions and replace them with more meaningful code - like ask a user to check if some golden image fits what he's seeing on the physical device screen.

File: src/ragger/navigator/interface.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        :type crop_first: Crop
+        :param crop_last: Crop (left, upper, right or lower pixels) last snapshot image for comparison
+                          (useful if using a generic snapshot).
+        :type crop_last: Crop
+
+        :return: img_idx
+        :rtype: int
+        """
+
+        if not isinstance(self._backend, SpeculosBackend):
```
Comment (lpascal-ledger) [inline src/ragger/navigator/interface.py:235]: Same remark as in previous method.

File: src/ragger/navigator/interface.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+            # When not using Speculos backend, taking snapshots is not possible
+            # therefore comparison is not possible too.
+            # TODO request user to interact with the device.
+            return 0
+
+        snaps_golden_path = self._get_snaps_dir_path(path, test_case_name, True)
+        if not snaps_golden_path.is_dir():
```
Comment (lpascal-ledger) [inline src/ragger/navigator/interface.py:242]: Same remark as before. It's reasonnable to ask `self._get_snaps_dir_path` to assert if the directory exists or not (and is a directory). It could be optionnable through a keyword argument if we don't want to check everytime.

File: src/ragger/navigator/interface.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+        # It saves time in non-nominal cases where the navigation flow does not start.
+        if self._compare_snap_with_timeout(first_golden_snap, timeout_s=2, crop=crop_first):
+            start = time()
+            # Navigate until the last snapshot specified in argument is found.
+            while not self._compare_snap_with_timeout(
+                    last_golden_snap, timeout_s=0.5, crop=crop_last):
+                now = time()
+                # Global navigation loop timeout in case the snapshot is never found.
+                if (now - start > timeout):
+                    raise ValueError(f"Timeout waiting for snap {last_golden_snap}")
```
Comment (lpascal-ledger) [inline src/ragger/navigator/interface.py:260]: A `TimeoutError` would be more appropriate.

File: src/ragger/navigator/interface.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+            # Validation action when last snapshot is found.
+            self.navigate([validation_instruction])
+
+            # Make sure there is a screen update after the final action.
+            start = time()
+            last_screen_update_timeout = 2
+            while self._compare_snap_with_timeout(last_golden_snap, timeout_s=0.5, crop=crop_last):
+                now = time()
+                if (now - start > last_screen_update_timeout):
+                    raise ValueError(
```
Comment (lpascal-ledger) [inline src/ragger/navigator/interface.py:283]: `TimeoutError`?

File: src/ragger/navigator/nano.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+from ragger.backend import BackendInterface
+from ragger.firmware import Firmware
+from ragger.navigator import NavigatorInterface, NavigationInstruction
+
+
+class NanoNavigator(NavigatorInterface):
+
+    _ARGS_KEY = 'args'
+
+    def __init__(self, backend: BackendInterface, firmware: Firmware):
```
Comment (lpascal-ledger) [inline src/ragger/navigator/nano.py:27]: I think this `__init__` is superfluous, as it just redirect the argument to it motherclass, in the exact same order.

File: src/ragger/navigator/nano.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+        :param instructions: Set of navigation instructions.
+        :type instructions: List[NavigationInstruction]
+
+        :raises NotImplementedError: If the navigation instruction is not implemented.
+
+        :return: None
+        :rtype: NoneType
+        """
+        for instruction in instructions:
```
Comment (lpascal-ledger) [inline src/ragger/navigator/nano.py:42]: One way to uncouple the data from the code flow, in order to avoid future very long `if ... elif ... else ...` structure with potential redundant conditions we could miss, would be to create a dictionnary `{ <instruction> : <callback> }` (or whatever structure), then have a code like:
```python
for instruction in instructions:
    if not instruction in CALLBACKS:
        raise NotImplementedError
    CALLBACKS[instruction]()
```

File: src/ragger/utils/misc.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+"""
+from typing import Optional, Tuple
+from pathlib import Path
+
+
+def app_path_from_app_name(app_dir, app_name: str, device: str) -> Path:
```
Comment (lpascal-ledger) [inline src/ragger/utils/misc.py:20]: `app_dir: Path`

**Review by lpascal-ledger — CHANGES_REQUESTED** [review]

File: src/ragger/backend/interface.py [hunk 18]
```diff
@@ -289,76 +289,11 @@ def both_click(self) -> None:
         raise NotImplementedError
 
     @abstractmethod
-    def navigate_and_compare_until_snap(self,
-                                        path: Path,
-                                        test_case_name: Path,
-                                        start_img_idx: int = 0,
-                                        last_img_idx: int = 0) -> bool:
+    def take_snapshot(self) -> Image:
```
[the remark this answers is not in the document]
Comment (lpascal-ledger) [reply src/ragger/backend/interface.py:292]: (or `io.BytesIO`)

File: src/ragger/navigator/nano.py (hunk 16 above)
Comment (lpascal-ledger) [reply src/ragger/navigator/nano.py:42]: (this could be onlined with a raison `lambda` but it would be very unclear)

Comment (lpascal-ledger) [reply src/ragger/navigator/nano.py:42]: Advantage of having a dedicated `{ <instruction>: <callback> }` structure is that it is very easy to register (add/modify) custom behavior on any instruction.

File: src/ragger/backend/interface.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (xchapron-ledger) [reply src/ragger/backend/interface.py:292]: Ok, I'm changing to io.BytesIO

File: src/ragger/navigator/nano.py (hunk 16 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/nano.py:43]: What if the instruction needs parameters? Let's say the `WAIT_1SEC` I've just added calls time.sleep(1). Shoud I create a `def sleep1(): time.sleep(1)`?

Comment (xchapron-ledger) [reply src/ragger/navigator/nano.py:43]: Or maybe can you think of a way to add parameters? Maybe instructions could be not just an enum, but a class containing an enum and args?

File: src/ragger/navigator/interface.py [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+    def _get_snaps_dir_path(self, path: Path, test_case_name: Path, is_golden: bool) -> Path:
+        if is_golden:
+            subdir = "snapshots"
+        else:
+            subdir = "snapshots-tmp"
+        return path / subdir / self._firmware.device / test_case_name
+
+    def _init_snaps_temp_dir(self, path: Path, test_case_name: Path) -> Path:
+        snaps_tmp_path = self._get_snaps_dir_path(path, test_case_name, False)
+        if snaps_tmp_path.exists():
```
[the remark this answers is not in the document]
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py:66]: it will raise an error I guess. I don't think this is an issue as I don't think it will happen? And If not, the error should be quite understandable?

File: README.md [hunk 20]
```diff
@@ -79,68 +83,161 @@ following fixtures:
 
 import pytest
```
Comment (fbeutin-ledger) [inline README.md:84]: Maybe put the conftest template in a dedicated file and refer to it in the README ?
Just a suggestion, I'm not even sure it would be better

File: README.md [hunk 21]
```diff
@@ -79,68 +83,161 @@ following fixtures:
...
+
+@pytest.fixture(scope="session")
+def display(pytestconfig):
+    return pytestconfig.getoption("display")
+
+
 @pytest.fixture
-def firmware():
-    return NANOS_FIRMWARE
+def test_name(request):
```
Comment (fbeutin-ledger) [inline README.md:122]: Maybe make it a property ?

Even make it two properties:
self.test_name => return request.node.name
self.test_name_no_suffix => return self.test_name.split("[")[0]

File: src/ragger/navigator/interface.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/ragger/navigator/interface.py:66]: `snaps_tmp_path.isdir()` will ensure it exists + is a directory

File: src/ragger/navigator/interface.py (hunk 10 above)
Comment (fbeutin-ledger) [reply src/ragger/navigator/interface.py:164]: +1
And Navigation DOES make sense in LedgerComm and LedgerWallet, even if it's just asking the tester `Press right button once` and `Ensure that this image corresponds to the current Nano screen`

File: src/ragger/navigator/interface.py [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+        """
+        if not isinstance(self._backend, SpeculosBackend):
+            # When not using Speculos backend, taking snapshots is not possible
+            # therefore comparison is not possible too and eventually this method
+            # fallback to navigate one.
+            return self.navigate(instructions)
+
+        snaps_tmp_path = self._init_snaps_temp_dir(path, test_case_name)
+
+        sleep(first_instruction_wait)
```
Comment (fbeutin-ledger) [inline src/ragger/navigator/interface.py:172]: Perfectly fine for a first step but it would be nice to eventually wait an event from speculos stating it is finished displaying
Maybe add a todo ?

File: src/ragger/utils/__init__.py [hunk 23]
```diff
@@ -15,5 +15,11 @@
...
 from .packing import pack_APDU
+from .path import BtcDerivationPathFormat, pack_derivation_path
+from .path import bitcoin_pack_derivation_path
+from .misc import app_path_from_app_name, prefix_with_len
+from .misc import create_currency_config
 
-__all__ = ("RAPDU", "pack_APDU", "Crop")
+__all__ = ("RAPDU", "pack_APDU", "Crop", "BtcDerivationPathFormat", "pack_derivation_path",
+           "bitcoin_pack_derivation_path", "app_path_from_app_name", "prefix_with_len",
+           "create_currency_config")
```
Comment (fbeutin-ledger) [inline src/ragger/utils/__init__.py:25]: Ok this effectively make the individual utils files redundant, good

File: src/ragger/navigator/nano.py (hunk 16 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/nano.py]: Done!

File: src/ragger/navigator/interface.py (hunk 22 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py:182]: yeah, could be nice to catch events for this...
I will add a todo for now.

File: README.md (hunk 20 above)
Comment (xchapron-ledger) [reply README.md]: Done

File: src/ragger/navigator/interface.py [hunk 24]
```diff
@@ -0,0 +1,288 @@
...
+            return screenshot_equal(f"{path}", snap)
+
+    def _compare_snap_with_timeout(self,
+                                   path: Path,
+                                   timeout_s: float = 5.0,
+                                   crop: Crop = None) -> bool:
+        start = time()
+        now = start
+        while not (now - start > timeout_s):
+            snap = self._backend.take_snapshot()
```
[the remark this answers is not in the document]
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py:106]: `BackendInterface.take_snapshot` returns a `BytesIO` in the current version (I changed it I don't know exactly when).

File: src/ragger/navigator/interface.py [hunk 25]
```diff
@@ -0,0 +1,288 @@
...
+        if not snaps_golden_path.is_dir():
+            raise ValueError(f"Golden snapshots directory ({snaps_golden_path}) does not exist.")
+
+        if not snaps_tmp_path.is_dir():
+            raise ValueError(f"Temporary snapshots directory ({snaps_golden_path}) does not exist.")
+
+        for i in range(0, last_img_idx + 1):
+            golden = self._get_snap_path(snaps_golden_path, i)
+            tmp = self._get_snap_path(snaps_tmp_path, i)
+            if not screenshot_equal(str(golden), str(tmp)):
```
[the remark this answers is not in the document]
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py:125]: it is located in Speculos, so I don't want to change this.
But it already handles either path as str and BytesIO.

File: src/ragger/navigator/interface.py (hunk 10 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py:174]: OR !

As discussed with @agrojean-ledger we could:
- add a `parser.addoption("--golden_snapshot")`
- pass it in `SpeculosBackend` instantiation `kwargs`
- have a specific behavior in theses cases which instead of comparing the snapshot it saves it in the golden dir.

with such behavior, we could go back to comparing the snapshot after each instruction, which would allow to do it on `LedgerComm` and `LedgerWallet`.

Comment (fbeutin-ledger) [reply src/ragger/navigator/interface.py:174]: I suggest we postpone this to another PR, we'll see along the way

File: src/ragger/navigator/interface.py (hunk 11 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py:235]: Discussed above

File: src/ragger/navigator/interface.py (hunk 10 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py]: I've update the PR with an additional commit on which we can hopefully find a consensus!

File: src/ragger/navigator/interface.py [hunk 26]
```diff
@@ -0,0 +1,297 @@
...
+
+class NavIns:
+
+    def __init__(self, id: NavInsID, args=(), kwargs: Dict[str, Any] = {}):
+        self.id = id
+        self.args = args
+        self.kwargs = kwargs
+
+
+class NavigatorInterface(ABC):
```
Comment (agrojean-ledger) [inline src/ragger/navigator/interface.py:44]: @xchapron-ledger I think it shouldn't be called "Interface" as this class implements stuff. It's just a parent class to a device navigator... So you might just call it "Navigator" and rename the file `interface.py` to `navigator.py`

File: src/ragger/navigator/interface.py [hunk 27]
```diff
@@ -0,0 +1,297 @@
...
+                if (now - start > last_screen_update_timeout):
+                    raise TimeoutError(
+                        f"Timeout waiting for screen change after last snapshot : {last_golden_snap}"
+                    )
+        else:
+            raise ValueError(f"Could not find first snapshot {first_golden_snap}")
+        return img_idx
+
+
+class NanoNavigator(NavigatorInterface):
```
Comment (agrojean-ledger) [inline src/ragger/navigator/interface.py:288]: I would have had a file per device child class

File: src/ragger/navigator/interface.py (hunk 26 above)
Comment (xchapron-ledger) [reply src/ragger/navigator/interface.py]: done

**Review by lpascal-ledger — APPROVED** [review]

**Review by lpascal-ledger — APPROVED** [review]
